flang/lib/Lower/ConvertExpr.cpp

```diff
@@ -3317,6 +3317,39 @@ class ArrayExprLowering {
     return genarr(asScalarRef(sym));
   }
 
+  /// Base the indices according to the origin of an entity. This is needed for
+  /// operations like fir.array_coor that do not expect zero based indices.
+  llvm::SmallVector<mlir::Value>
+  applyShift(mlir::Value shape, mlir::Value slice,
+             llvm::ArrayRef<mlir::Value> indices) {
+    if (indices.empty())
+      return {};
+    auto type = indices[0].getType();
+    auto loc = getLoc();
+    if (!shape || slice || shape.getType().isa<fir::ShapeType>()) {
+      // Entity lower bounds are one.
+      auto one = builder.createIntegerConstant(loc, type, 1);
+      llvm::SmallVector<mlir::Value> newIndices;
+      for (auto i : indices)
+        newIndices.push_back(builder.create<mlir::AddIOp>(loc, i, one));
```

Comment:
Do you want to add a `createConvert(one)` here to be on the safe side?

Reply:
Shared `originateIndices` instead.